### PR TITLE
Add assertions related to interconnects

### LINKED ISSUE #164: System Interconnection Table Completed
**Extended Description**
- Description...

**Preconditions**
- Preconditions...

**Acceptance Critera**
- Acceptance criteria...

**Story Tasks**
- [ ] Tasks...

**Definition of Done**
 - [ ] Acceptance criteria met - Each user story should meet the acceptance criteria in the description
 - [ ] Unit test coverage of our code > 90%  (*from [QASP](https://github.com/flexion/TTS-10x/wiki/TTS-QASP)*) this may be fuzzy and hard to prove
 - [ ] Code quality checks passed - Enable html tidy with XML code standards as part of the build (*from [QASP](https://github.com/flexion/TTS-10x/wiki/TTS-QASP)*)
 - [ ] Accessibility: (*from [QASP](https://github.com/flexion/TTS-10x/wiki/TTS-QASP)*) as we create guidance or documentation and reports (semantic tagging including aria tags): demonstrate with 0 errors reported for WCAG 2.1 AA standards using an automated scanner and 0 errors reported in manual testing
 - [ ] Code reviewed - Code reviewed by at least one other team members (or developed by a pair)
 - [ ] Source code merged - Code that’s demoed must be in source control and merged
 - [ ] Code must successfully build and deploy into staging environment (*from [QASP](https://github.com/flexion/TTS-10x/wiki/TTS-QASP)*): this may evolve from xslt sh pipline into something more
 - [ ] Security reviewed and reported - Conduct vulnerability and compliance scanning. threat modeling? 
 - [ ] Code submitted must be free of medium- and high-level static and dynamic security vulnerabilities (*from [QASP](https://github.com/flexion/TTS-10x/wiki/TTS-QASP)*)
 - [ ] Usability tests passed - Each user story should be easy to use by target users (development community? FedRAMP FART team)
 - [ ] Usability testing and other user research methods must be conducted at regular intervals throughout the development process (not just at the beginning or end). (*from [QASP](https://github.com/flexion/TTS-10x/wiki/TTS-QASP)*)
 - [ ] Code refactored for clarity - Code must be clean, self-documenting
 - [ ] No local design debt
 - [ ] Load/performance tests passed - test data needed - saxon instrumentation
 - [ ] Documentation generated - update readme or contributing markdown as necessary.
 - [ ] Architectural Decision Record completed as necessary for significant design choices

### PR DESCRIPTION
**Pull request by GaryGapinski**
This PR adds assertions related to interconnections.

Closes #164.

### CODE REVIEW

File: src/validations/rules/ssp.sch [hunk 1]
```diff
@@ -4059,5 +4335,149 @@
...
+        <sch:diagnostic
+            doc:assertion="interconnection-has-allowed-interconnection-security-value"
+            doc:context="oscal:component[@type = 'interconnection']/oscal:prop[@ns eq 'https://fedramp.gov/ns/oscal' and @name eq 'interconnection-security']"
+            id="interconnection-has-allowed-interconnection-security-value-diagnostic">A system interconnection lacks an allowed
+            interconnection-security.</sch:diagnostic>
+        <sch:diagnostic
+            doc:assertion="interconnection-has-allowed-interconnection-security-remarks"
+            doc:context="oscal:component[@type = 'interconnection']/oscal:prop[@ns eq 'https://fedramp.gov/ns/oscal' and @name eq 'interconnection-security']"
+            id="interconnection-has-allowed-interconnection-security-remarks-diagnostic">This system interconnection with an interconnection-security
+            of &quot;other&quot; lacks explanatory remarks.</sch:diagnostic>
```
Comment (ohsh6o) [inline src/validations/rules/ssp.sch:4352]:
```suggestion
            id="interconnection-has-allowed-interconnection-security-remarks-diagnostic">This system interconnection defines an alternate method for securing information in transit, where interconnection-security
            is defined as &quot;other&quot; and the required explanatory remarks are missing.</sch:diagnostic>
```

File: src/validations/test/ssp.xspec [hunk 2]
```diff
@@ -9075,4 +9075,952 @@
             </x:scenario>
         </x:scenario>
     </x:scenario>
+
+    <x:scenario
+        label="FEDRAMP OSCAL SSP Interconnections">
```
Comment (ohsh6o) [inline src/validations/test/ssp.xspec:9080]:
```suggestion
        label="FedRAMP OSCAL SSP Interconnections">
```

**Review by ohsh6o — COMMENTED** [review]
I am not sure if I placed my wording recommendations correctly. Should I have been focusing on the `diagnostic`s?

File: src/validations/rules/ssp.sch [hunk 3]
```diff
@@ -2942,6 +2950,274 @@
...
+                test="oscal:prop[@name eq 'ipv4-address' and @class eq 'remote']">A system interconnection must specify the external system IPv4
+                address.</sch:assert>
+            <sch:assert
+                diagnostics="interconnection-has-remote-IPv6-address-diagnostic"
+                doc:guide-reference="DRAFT Guide to OSCAL-based FedRAMP System Security Plans §4.20"
+                doc:template-reference="System Security Plan Template §11"
+                id="interconnection-has-remote-IPv6-address"
+                role="error"
+                test="oscal:prop[@name eq 'ipv6-address' and @class eq 'remote']">A system interconnection must specify the external system IPv6
+                address.</sch:assert>
```
[the remark this answers is not in the document]
Comment (GaryGapinski) [reply src/validations/rules/ssp.sch]: There is no (current) attribute of a system interconnection from which an endpoint's IPv4, or IPv6, or both, address necessity can be derived. The current assertions demand both.

[the remark this answers is not in the document]
Comment (ohsh6o) [reply src/validations/rules/ssp.sch]: Let's discuss that today then.

File: src/validations/rules/ssp.sch [hunk 4]
```diff
@@ -2942,6 +2950,274 @@
...
+
+        <sch:rule
+            context="oscal:component[@type eq 'interconnection']/oscal:protocol/oscal:port-range">
+            <sch:assert
+                diagnostics="interconnection-protocol-port-range-has-transport-diagnostic"
+                doc:guide-reference="DRAFT Guide to OSCAL-based FedRAMP System Security Plans §4.20"
+                doc:template-reference="System Security Plan Template §11"
+                id="interconnection-protocol-port-range-has-transport"
+                role="error"
+                test="@transport">A system interconnection protocol port range declaration must state a transport protocol.</sch:assert>
```
[the remark this answers is not in the document]
Comment (GaryGapinski) [reply src/validations/rules/ssp.sch:3203]: The OSCAL values are insufficient. This should be unconstrained unless/until a suitable set of allowed values is defined.

File: src/validations/rules/ssp.sch (hunk 3 above)
[the remark this answers is not in the document]
Comment (GaryGapinski) [reply src/validations/rules/ssp.sch]: Assertion now requires paired IPv4 or IPv6 addresses.

**Review by ohsh6o — APPROVED** [review]

**Pull request merged**